src/app/routes/repo.rs

```diff
@@ -0,0 +1,131 @@
+use crate::{
+    app::auth::middleware::SessionInfo,
+    service::database::{
+        conn::DbPool,
+        models::{Challenge, Repository, User},
+    },
+    shared::errors::{CreateRepositoryError, RepositoryError},
+};
+use actix_web::{web, HttpMessage, HttpRequest, HttpResponse, Scope};
+use log::error;
+use reqwest::StatusCode;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CreateRepoResponse {
+    repo_name: String,
+    repo_url: String,
+}
+
+#[derive(Deserialize)]
+pub struct CreateRepoRequest {
+    repo_url: String,
+}
+
+pub fn init() -> Scope {
+    web::scope("/repo").route("", web::post().to(create_repo))
+}
+
+async fn create_repo(
+    req: HttpRequest,
+    body: Result<web::Json<CreateRepoRequest>, actix_web::Error>,
+    pool: web::Data<DbPool>,
+) -> Result<HttpResponse, RepositoryError> {
+    let repo_url = match body {
+        Ok(body) => body.repo_url.clone(),
+        Err(e) => {
+            error!("Error getting repository url: {}", e);
+            return Err(RepositoryError::BadRequest(
+                "Repository url is required".to_string(),
+            ));
+        }
+    };
+    if repo_url.is_empty() {
+        return Err(RepositoryError::BadRequest(
+            "Repository url is required".to_string(),
+        ));
+    }
+
+    let client = reqwest::Client::new();
+    let git_service_url = std::env::var("GIT_SERVICE_URL").expect("GIT_SERVICE_URL not set");
+    let mut conn = pool.get().expect("couldn't get db connection from pool");
+
+    let user_id = match req.extensions().get::<SessionInfo>() {
+        Some(session_info) => session_info.user_id,
+        None => {
+            return Err(RepositoryError::BadRequest(
+                "User not authenticated".to_string(),
+            ));
+        }
+    };
+    let user = match User::get_user(&mut conn, Some(&user_id), None, None, None) {
+        Ok(user) => user,
+        Err(e) => {
+            error!("Error getting user: {}", e);
+            return Err(RepositoryError::BadRequest("User not found".to_string()));
+        }
+    };
+
+    let challenge = match Challenge::get_challenge(&mut conn, None, Some(&repo_url)) {
+        Ok(challenge) => challenge,
+        Err(e) => {
+            error!("Error getting challenge for repository url: {}", e);
+            return Err(RepositoryError::BadRequest(format!(
+                "repository url starter code {} not found",
+                &repo_url
+            )));
+        }
+    };
+
+    let repo_name = repo_url
+        .split("/")
+        .last()
+        .ok_or(RepositoryError::BadRequest(
+            "Error getting repository name from url".to_string(),
+        ))?;
```

Comment:
```suggestion
        .rsplit('/')
        .next()
        .and_then(|name| if name.is_empty() { None } else { Some(name) })
        .ok_or_else(|| {
            error!("Invalid repository URL format: {}", repo_url);
            RepositoryError::BadRequest("Invalid repository URL format".to_string())
        })?;
```

We should validate the format of the url before working on the string to ensure we get expected behaviour.

Reply:
addressed https://github.com/Extheoisah/hxckr-core/pull/19/commits/3f24e614b071801e3ee2d11924ee8e6da32f700c